### PR TITLE
Feature/email confirm

### PR DESCRIPTION
**Pull request by dawranliou**

### CODE REVIEW

File: accounts/tests.py [hunk 1]
```diff
@@ -156,3 +159,46 @@ def test_user_refresh_token(self):
...
+            'password': 'testpassword',
+            'email': 'test@example.com'
+        }
+        self.signup_url = reverse('account_create')
+
+    def test_new_signup_account_not_activated(self):
+        signup_response = self.client.post(self.signup_url, self.user_credential, format='json')
+
+        self.assertEqual(signup_response.status_code, status.HTTP_201_CREATED)
+        self.assertFalse(User.objects.last().is_active)
```
Comment (dawranliou) [inline accounts/tests.py:179]: New signups should not have users activated.

File: accounts/tests.py [hunk 2]
```diff
@@ -156,3 +159,46 @@ def test_user_refresh_token(self):
...
+
+    def test_email_confirmation_is_sent(self):
+        self.client.post(self.signup_url, self.user_credential, format='json')
+
+        user = User.objects.last()
+        self.assertEqual(len(mail.outbox), 1)
+        confirmation_email = mail.outbox[0]
+        self.assertIn(user.username, confirmation_email.body)
+        urls = SignupTest.find_links(confirmation_email.body)
+        self.assertTrue(urls)
```
Comment (dawranliou) [inline accounts/tests.py:189]: Link(s) should be present in the confirmation email body.

File: accounts/tests.py [hunk 3]
```diff
@@ -156,3 +159,46 @@ def test_user_refresh_token(self):
...
+
+    def test_activate_account(self):
+        self.client.post(self.signup_url, self.user_credential, format='json')
+
+        confirmation_email = mail.outbox[0]
+        urls = SignupTest.find_links(confirmation_email.body)
+        for url in urls:
+            self.client.get(url)
```
Comment (dawranliou) [inline accounts/tests.py:197]: Make GET request(s) to hit the url(s).

File: accounts/tests.py [hunk 4]
```diff
@@ -156,3 +159,46 @@ def test_user_refresh_token(self):
...
+
+        user = User.objects.last()
+        self.assertEqual(user.is_active, True)
```
Comment (dawranliou) [inline accounts/tests.py:200]: User should be activated by now.

File: accounts/tests.py [hunk 5]
```diff
@@ -156,3 +159,46 @@ def test_user_refresh_token(self):
         refresh_response = self.client.post(self.token_refresh_url, self.user_credential, format='json')
 
         self.assertNotEqual(token, refresh_response.data['token'])
+
+
+class SignupTest(APITestCase):
+
+    def setUp(self):
```
Comment (cindycaobutler) [inline accounts/tests.py:167]: If this is your full unittest, please cover a few invalid username/password cases, e.g. empty username, username too long(https://www.google.com/search?q=username+length&oq=username+length&aqs=chrome..69i57j0l5.4219j0j7&sourceid=chrome&ie=UTF-8), password should meet policies (https://en.wikipedia.org/wiki/Password_policy). Our DB will have constraints but we don't want to always insert invalid username/password and then let DB throw error. Our backend service should immediately return error which is faster.

File: accounts/views.py [hunk 6]
```diff
@@ -20,7 +26,49 @@ def post(self, request, format='json'):
...
+                send_mail('Confirm your email', message, from_email='localhost', recipient_list=[user.email])
                 return Response(serializer.data, status.HTTP_201_CREATED)
 
         return Response(serializer.errors, status.HTTP_400_BAD_REQUEST)
+
+
+class ActivateAccount(APIView):
+    """API to activate account after signing-up.
+
+    get:
```
Comment (cindycaobutler) [inline accounts/views.py:55]: Please use named type annotation in input parameters, e.g. (input_str: str)

Please follow this style for comments:

"""<function purpose>

Args:
    <arg1>: description
    <arg2>: description

Returns:
    <return type>: description
"""

**Review by cindycaobutler — CHANGES_REQUESTED** [review]
A global comment to all functions:

Please use named type annotation in input parameters, e.g. (input_str: str)

Please follow this style for comments:

"""<function purpose>

Args:
    <arg1>: description
    <arg2>: description

Returns:
    <return type>: description
"""

File: accounts/tests.py (hunk 5 above)
Comment (dawranliou) [reply accounts/tests.py:167]: The username and password validations are tested in the test case class - AccountsTest, which I believe already satisfied the cases you mentioned. I agree with you most validation shouldn't hit the DB. Basically only the uniqueness validation needs to hit the DB; others are handled by the request data serializer (see `accounts/serializers.py`.) The complete list of tests are:

1. test_create_user_return_valid_token
1. test_create_user_with_short_password
1. test_create_user_with_no_password
1. test_create_user_with_too_long_username
1. test_create_user_with_no_username
1. test_create_user_with_preexist_username - hits DB
1. test_create_user_with_preexist_email - hits DB
1. test_create_user_with_invalid_email
1. test_create_user_with_no_email

**Comment by dawranliou** [comment]
@cindycaobutler I committed the type annotations and the docstrings. Please review the changes. Thanks!

**Pull request merged**